kolibri/content/api.py

```diff
@@ -103,26 +104,55 @@ def filter_search(self, queryset, name, value):
         Implement various filtering strategies in order to get a wide range of search results.
         """
 
-        # search the whole string in the title or description
-        query = Q(title__icontains=value) | Q(description__icontains=value)
-
-        # split string up by spaces and search each individual string in title or description
-        str_arr = value.split()
-        for string in str_arr:
-            query.add(Q(title__icontains=string) | Q(description__icontains=string), Q.OR)
-        matches = queryset.filter(query)
-
-        # remove duplicate content items based on content_id
-        content_id_set = set()
-        id_list = []
-        for id_pair in matches.values('content_id', 'id'):
-            if id_pair['content_id'] in content_id_set:
-                pass
-            else:
-                content_id_set.add(id_pair['content_id'])
-                id_list.append(id_pair['id'])
-
-        return matches.filter(id__in=id_list[:30])
+        # return the result of and-ing a list of queries
+        def intersection(queries):
+            return reduce(lambda x, y: x & y, queries)
+
+        # return the result of or-ing a list of queries
+        def union(queries):
+            return reduce(lambda x, y: x | y, queries)
+
+        # queries ordered by relevance priority
+        words = value.split()
+        all_queries = [
+            # exact match in title
+            Q(title__icontains=value),
+            # all words in title
+            intersection([Q(title__icontains=w) for w in words]),
+            # exact match in description
+            Q(description__icontains=value),
+            # all words in description
+            intersection([Q(description__icontains=w) for w in words]),
+            # any stemmed words in title
+            union([Q(title__icontains=stemmer.stem(w)) for w in words]),
```

Comment:
Note: we stem the words in the last strategies as a last ditch effort to get some relevant results. Without the stemming, something like "sciences" would not get any hits.